compile-time-programming/requires-expressions.md

```diff
@@ -0,0 +1,102 @@
+# Module name: Requires Expressions
+
+## Overview
+
+<table>
+  <thead>
+    <th>Level</th>
+    <th>Objectives</th>
+  </thead>
+  <tr>
+    <td>Foundational knowledge</td>
+    <td>Define and use requires-expressions to check satisfaction of expressions by given parameters</td>
+  </tr>
+  <tr>
+    <td>Advanced</td>
+    <td>Define and use requires-expressions to check properties of expressions</td>
+  </tr>
+  <tr>
+    <td>Further studies</td>
+    <td></td>
+  </tr>
+</table>
+
+## Motivation
+
+Requires-expressions allow a developer to perform compile-time evaluation on the validity of other expressions. These are fundamental to the ability to write concepts.[[Compile-time programming: concepts]][1]
+
+## Topic introduction
+
+Requires-expressions are compile-time predicates which evaluate to true when their specified set of expressions are all valid for a given set of inputs.
+
+## Foundational knowledge: Writing requires-expressions
+
+### Background/Required Knowledge
+
+A student is able to:
+
+* Write and use a function template [[Compile-time programming: function templates]][2]
+* Differentiate between expressions and statements
+
+It is helpful if:
+
+* A student is aware of the constraints imposed by their template definitions
+
+### Student outcomes
+
+A student should be able to:
+
+1. Write a simple-requirement to assert the validity of an expression
+2. Write a type-requirement to check the existence of a type by its identifier
+3. Write a compound-requirement to test the resulting type of an expression
+4. Write a nested-requirement to test the constexpr value of an operation, as opposed to just the semantic validity
+5. Use a requires-expression within a concept, requires-clause, or `if constexpr` condition
+
+### Caveats
+
+To require expressions which evaluate to a boolean value, such as `sizeof(t) == 4`, to only be satisfied when the result of the expression is `true`, you must use a nested-requirement, such as `requires sizeof(t) == 4;`. Omitting the `requires` results in a simple-requirement, which is satisfied based purely on semantic validity, not on the result of the operation.
```

Comment:
Ooh ooh! If we're doing pitfalls, see my talk "Concepts As She Is Spoke" (2018). I've got tons. https://godbolt.org/z/aLyaZb

        template<class T>
        concept False = requires(T t) {
            t.value -> size_t;
        };
        struct S { size_t value; };
        static_assert(!False<S>);

Reply:
Oh, and again you said `semantic` when you meant `syntactic`.

Reply:
That's certainly an interesting example. Had to play around with it for a while, before I found a way to express it. I'll add a point-to-cover on the value of adding static_asserts to both positively and negatively test the concept for known types

Reply:
just realized that's more of a check to be performed when using a concept, which I am trying to not imply as required to use requires-expressions

Reply:
> Ooh ooh! If we're doing pitfalls, see my talk "Concepts As She Is Spoke" (2018). I've got tons. https://godbolt.org/z/aLyaZb
> 
> ```
>     template<class T>
>     concept False = requires(T t) {
>         t.value -> size_t;
>     };
>     struct S { size_t value; };
>     static_assert(!False<S>);
> ```

This won't compile because `t.value -> size_t` isn't valid syntax, so I'm not sure why it's a pitfall?

Reply:
:facepalm: I see why it's a pitfall now. Ignore me.

Reply:
I'm happy to add the example, but there's a lot to decompress there, and I'm not sure how to succinctly summarize it.

While I'm at it, just how _does_ one properly write a compound-requirement where you would want to remove cv-ref? I can get it via a nested requirement with a decltype and decay_t, but the compound version eludes me...
https://godbolt.org/z/GKrf3j

Reply:
@rwdougla, re your nerdsnipe: There's no built-in syntax for that. I would do it this way: https://godbolt.org/z/vd8Wdr

    template<class T, class U>
    concept decays_to = std::same_as<std::decay_t<T>, U>;
    ... { t.value } -> decays_to<size_t>; ...
(There's an active C++23 proposal for "built-in _DECAY_COPY_" which would make `{ auto(t.value) } -> std::same_as<size_t> }` also work in this case, but only because our target type `size_t` happens to be copy-constructible.)

Come to think of it, this would also work: https://godbolt.org/z/hKW661

    ... { std::as_const(t.value) } -> std::same_as<const size_t&>; ...
But I _really_ wouldn't recommend that. It requires the programmer to think inside-out from what they're presumably really trying to do. (Which is maybe the real lesson here. Are you sure you _want_ to write such a requirement? Why? What does it actually let you say about the type `T` that will be useful to you in a template definition?)

Reply:
And since this whole cvref thing seems like maybe an XY problem, let me just rewind and be explicit about the original example, for those who can't find the time to watch the CppCon talk:

    { t.value } -> std::convertible_to<size_t>;  // Well-formed

This says that `decltype((t.value))` ought to be convertible to `size_t`. Awesome. Unsurprising. Just a _soupçon_ of implementation divergence between Clang [wrong] and MSVC/GCC [right] about the proper number of parens in `decltype((t.value))`.

    { t.value } -> size_t;  // Ill-formed, diagnostic required

This says that `decltype((t.value))` ought to satisfy `concept size_t`. It's overwhelmingly likely that no such concept is in scope, and so this requirement would be flagged as a hard error and your program wouldn't ever compile.

    t.value -> size_t;  // Well-formed and SILENTLY WRONG

This says that the syntactically valid expression `t.value->size_t` ought to be well-formed (and we don't care about what type it is). It's overwhelmingly likely that this expression is _not_ well-formed for your particular `T`, so `False<T>` silently evaluates to `false`.